src/desul/array.hpp

```diff
@@ -0,0 +1,301 @@
+         static_assert(N > 0, "Calling desul::array::back on an empty array is not allowed.");
+         return elements[size() - 1];
+      }
+
+      DESUL_HOST_DEVICE constexpr const_reference back() const {
+         static_assert(N > 0, "Calling desul::array::back on an empty array is not allowed.");
+         return elements[size() - 1];
+      }
+
+      DESUL_HOST_DEVICE constexpr pointer data() noexcept {
+         return elements;
+      }
+
+      DESUL_HOST_DEVICE constexpr const_pointer data() const noexcept {
+         return elements;
+      }
+
+      DESUL_HOST_DEVICE constexpr iterator begin() noexcept {
+         return elements;
+      }
+
+      DESUL_HOST_DEVICE constexpr const_iterator begin() const noexcept {
+         return elements;
+      }
+
+      DESUL_HOST_DEVICE constexpr const_iterator cbegin() const noexcept {
+         return elements;
+      }
+
+      DESUL_HOST_DEVICE constexpr iterator end() noexcept {
+         return &elements[size()];
+      }
+
+      DESUL_HOST_DEVICE constexpr const_iterator end() const noexcept {
+         return &elements[size()];
+      }
+
+      DESUL_HOST_DEVICE constexpr const_iterator cend() const noexcept {
+         return &elements[size()];
+      }
+
+      DESUL_HOST_DEVICE constexpr bool empty() const noexcept {
+         return size() == 0;
+      }
+
+      DESUL_HOST_DEVICE constexpr size_type size() const noexcept {
+         return N;
+      }
+
+      DESUL_HOST_DEVICE constexpr size_type max_size() const noexcept {
+         return size();
+      }
+
+      DESUL_HOST_DEVICE constexpr void fill(const T& value) {
+         for (std::size_t i = 0; i < size(); ++i) {
+            elements[i] = value;
+         }
+      }
+
+      DESUL_HOST_DEVICE constexpr void swap(array& other) noexcept(std::is_nothrow_swappable_v<T>) {
+         for (std::size_t i = 0; i < size(); ++i) {
+            desul::swap(elements[i], other[i]);
+         }
+      }
+
+      value_type elements[N];
+   };
+
+   template <class T, std::size_t N>
+   DESUL_HOST_DEVICE constexpr bool operator==(const array<T, N>& lhs,
+                                               const array<T, N>& rhs) {
+      for (std::size_t i = 0; i < N; ++i) {
+         if (lhs[i] != rhs[i]) {
+            return false;
+         }
+      }
+
+      return true;
+   }
+
+   template <class T, std::size_t N>
+   DESUL_HOST_DEVICE constexpr bool operator!=(const array<T, N>& lhs,
+                                               const array<T, N>& rhs) {
+      return !(lhs == rhs);
+   }
+
+   template <class T, std::size_t N>
+   DESUL_HOST_DEVICE constexpr bool operator<(const array<T, N>& lhs,
+                                              const array<T, N>& rhs) {
+      for (std::size_t i = 0; i < N; ++i) {
+         if (lhs[i] < rhs[i]) {
+            return true;
+         }
+         else if (lhs[i] == rhs[i]) {
+            continue;
+         }
+         else {
+            return false;
+         }
```

Comment:
Should this all be done in terms of operator<?
```suggestion
         else if ( rhs[i] < lhs[i]) {
            return false;
         }
```